sw/host/opentitantool/src/command/i2c.rs

```diff
@@ -105,6 +107,156 @@ impl CommandDispatch for I2cRawWriteRead {
     }
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq, ValueEnum)]
+pub enum Mode {
+    Host,
+    Device,
+}
+
+#[derive(Debug, Args)]
+pub struct I2cSetMode {
+    pub mode: Mode,
+
+    /// 7 bit I2C device address.
+    #[arg(
+        short,
+        long,
+        value_parser = u8::from_str
+    )]
+    pub addr: Option<u8>,
+}
+
+impl CommandDispatch for I2cSetMode {
+    fn run(
+        &self,
+        context: &dyn Any,
+        transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        transport.capabilities()?.request(Capability::I2C).ok()?;
+        let context = context.downcast_ref::<I2cCommand>().unwrap();
+        let i2c_bus = context.params.create(transport, "DEFAULT")?;
+        match self.mode {
+            Mode::Host => i2c_bus.set_mode_host()?,
+            Mode::Device => match self.addr {
+                Some(addr) => i2c_bus.set_mode_device(addr)?,
+                None => bail!("Must specify `--addr` after `set-mode device`"),
+            },
+        }
+        Ok(None)
+    }
+}
+
+/// Retrieve transcript of transfers since last time, and whether the I2C host is currently
+/// waiting to READ data.
+#[derive(Debug, Args)]
+pub struct I2cGetDeviceStatus {
+    /// For how long to wait, in case of nothing to report.
+    #[arg(short,
+    long,
+    default_value="500ms",
+    value_parser = humantime::parse_duration)]
+    timeout: Duration,
+}
+
+#[derive(Debug, serde::Serialize)]
+pub enum DeviceTransfer {
+    /// The I2C host read a number of previously prepared bytes.
+    Read {
+        addr: u8,
+        /// True if `I2cPrepareRead` was not called in time.
+        timeout: bool,
+        len: usize,
+    },
+    /// The I2C host wrote the given sequence of bytes.
+    Write { addr: u8, hexdata: String },
+}
+
+#[derive(Debug, serde::Serialize)]
+pub enum ReadStatus {
+    /// Host has asked to read data, debugger device is currently stretching the clock waiting to
+    /// be told what data to transmit via I2C.
+    WaitingForData,
+    /// Host is not asking to read data, and debugger also does not have any prepared data.
+    Idle,
+    /// Host is not asking to read data, debugger has prepared data for the case the host should
+    /// ask in the future.
+    DataPrepared,
+}
```

Comment:
Do these structs simply mirror their counterparts in the `opentitanlib::io::i2c` namespace?  Should we just make those `serde::Serialize` rather than duplicating them here?

Reply:
They almost, but do not quite mirror the structs from `io`.  This one has `hexdata` field for instance, and could in the future also maybe present 4-byte responses in a `uint32` field next to `hexdata`.

I think it makes sense to keep the `io` version with `Vec<u8>`, and then separate structs for presentation.